src/Psalm/Internal/Type/SimpleAssertionReconciler.php

```diff
@@ -2021,20 +2025,53 @@ private static function reconcileList(
         $did_remove_type = false;
 
         foreach ($existing_var_atomic_types as $type) {
-            if ($type instanceof TList
-                || ($type instanceof TKeyedArray && $type->is_list)
-            ) {
-                if ($is_non_empty && $type instanceof TList && !$type instanceof TNonEmptyList) {
+            if ($type instanceof TList) {
+                if ($is_non_empty && !$type instanceof TNonEmptyList) {
                     $array_types[] = new TNonEmptyList($type->type_param);
                     $did_remove_type = true;
                 } else {
                     $array_types[] = $type;
                 }
-            } elseif ($type instanceof TArray || $type instanceof TKeyedArray) {
-                if ($type instanceof TKeyedArray) {
-                    $type = $type->getGenericArrayType();
-                }
+            } elseif ($type instanceof TKeyedArray) {
+                if ($type->is_list) {
+                    $array_types[] = $type;
+                } else {
+                    $did_remove_type = true;
+                    $type = clone $type;
+                    $min_unset_list_key = 0; // Minimum list key not explicitly set
+                    ksort($type->properties);
+                    foreach ($type->properties as $prop_key => $prop_value) {
+                        if (!is_int($prop_key)) {
+                            if ($prop_value->possibly_undefined) {
+                                unset($type->properties[$prop_key]);
+                            } else {
+                                // Can't reconcile, type is removed
+                                continue 2;
+                            }
+                        } elseif ($prop_key === $min_unset_list_key) {
+                            ++$min_unset_list_key;
```

Comment:
Actually I was just thinking about this, if we do `array_is_list(...)` on `array{0: int, 1?: int, 2?: int, 3: int}` the `1` and `2` keys should no longer be `possibly_undefined`. I'll go fix that shortly.